File: coding-interview-tree-traversal-advanced/src/main/java/com/example/treetraversals/models/TreeNode.java
```java
package com.example.treetraversals.models;

/**
 * Represents a node in a binary tree.
 * Each node has an integer value, and references to its left and right children.
 */
public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;

    /**
     * Default constructor for an empty node.
     */
    public TreeNode() {}

    /**
     * Constructor for a node with a given value.
     * Left and right children are initialized to null.
     * @param val The integer value of the node.
     */
    public TreeNode(int val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

    /**
     * Constructor for a node with a given value and specific left and right children.
     * @param val The integer value of the node.
     * @param left The left child node.
     * @param right The right child node.
     */
    public TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    @Override
    public String toString() {
        return "TreeNode{" + "val=" + val + '}';
    }
}

```

File: coding-interview-tree-traversal-advanced/src/data-structures/TreeNode.js
```javascript
/**
 * @fileoverview Defines the TreeNode class for binary tree implementation.
 */

/**
 * Represents a node in a Binary Tree.
 * Each node has a value, and references to its left and right children.
 */
class TreeNode {
    /**
     * Constructs a new TreeNode.
     * @param {*} val The value to store in the node. Defaults to 0.
     * @param {TreeNode|null} left The left child node. Defaults to null.
     * @param {TreeNode|null} right The right child node. Defaults to null.
     */
    constructor(val = 0, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    /**
     * Checks if the node is a leaf node (has no children).
     * @returns {boolean} True if it's a leaf, false otherwise.
     */
    isLeaf() {
        return this.left === null && this.right === null;
    }

    /**
     * Returns a string representation of the node.
     * @returns {string} The value of the node as a string.
     */
    toString() {
        return String(this.val);
    }
}

module.exports = TreeNode;
```